### PR TITLE
dts: add SPI device support

### PR DESCRIPTION
**Pull request by ydamigos**
This PR provides for SPI devices in device tree.

Adds generic yaml description for spi-devices such as sensors or bluetooth modules.
Provides dts yaml bindings for bluetooth modules used by 96b_carbon and disco_l475_iot1 boards.
Moves bluetooth module SPI port related configuration to DTS and generates dts based defines on 96b_carbon and disco_l475_iot1.

### CODE REVIEW

**Review by superna9999 — APPROVED** [review]
Ok for me

**Review by tbursztyka — APPROVED** [review]
Seems fine to me.

It looks like the yaml file you created for stm32 could be  generalized for all the other spi drivers (the ones that follow the current api, not the legacy one).

File: dts/bindings/spi/spi-device.yaml
Comment (erwango) [inline dts/bindings/spi/spi-device.yaml]: Not sure this is right for all SPI modules.
SPBTLE uses GPIO chip select and hence reg has provided here has no meaning.
So, I guess we should keep it (if valid for some modules) but should be optional

File: boards/arm/disco_l475_iot1/disco_l475_iot1.dts
Comment (erwango) [inline boards/arm/disco_l475_iot1/disco_l475_iot1.dts]: Remove this line

**Review by erwango — CHANGES_REQUESTED** [review]
One remark on SPI chip select otherwise LGTM!

File: dts/bindings/spi/spi-device.yaml
Comment (ydamigos) [reply dts/bindings/spi/spi-device.yaml]: I used the same approach as linux/Documentation/devicetree/bindings/spi/spi-bus.txt. I plan to add a property cs-gpios in spi.yaml when the PR for the gpio nodes is merged.
If, for example, the SPI has 2 CS lines, and the cs-gpios property looks like this:
```
cs-gpios = <&gpioA 12 0> <&gpioC 11 0>;
```
we have the following mapping:
```
cs0 : &gpioA 12 0
cs1 : &gpioC 11 0
```
The **0** in spi-device@**0** and property reg = <**0**> point to **cs0**.

Comment (erwango) [reply dts/bindings/spi/spi-device.yaml]: For once, I gave up waiting for my stalling editor to open my linux kernel repo... I should be more patient!
Thanks for the head's up, removing my comment.

**Review by erwango — APPROVED** [review]

File: dts/bindings/spi/spi-device.yaml
Comment (ydamigos) [reply dts/bindings/spi/spi-device.yaml]: I always use http://elixir.free-electrons.com/linux/latest/source to browse linux kernel.

**Comment by ydamigos** [comment]
Now that PR #5582 was merged, we can merge this one when it passes the review process.

File: dts/bindings/spi/spi-device.yaml
Comment (mbolivar) [inline dts/bindings/spi/spi-device.yaml]: This should be "optional", no? Not all SPI devices support CS pins

**Review by mbolivar — APPROVED** [review]
Seems fine, I have a question that's probably just me misunderstanding.

(I'm assuming this was tested on Carbon?)

File: dts/bindings/spi/spi-device.yaml
Comment (ydamigos) [reply dts/bindings/spi/spi-device.yaml]: I haven't come across a SPI device without CS pin. I would suggest to keep it as required since SPI is a bus. In case of a device without CS pin, the SPI bus would have one device on "address" reg = <0>.

**Comment by ydamigos** [comment]
@mbolivar Did you try it and didn't work? I tested it both on 96b_carbon and disco_l475_iot1.

File: dts/bindings/spi/spi-device.yaml
Comment (erwango) [reply dts/bindings/spi/spi-device.yaml]: > I plan to add a property cs-gpios in spi.yaml when the PR for the gpio nodes is merged.

Looking forward to see all bricks working together! This will require some extra modifications in extract_dts_includes.py to be fully functional (specially if we also add stuff like reset-gpios), but should not be to much of a burden

File: dts/bindings/spi/spi-device.yaml
Comment (mbolivar) [reply dts/bindings/spi/spi-device.yaml]: Lpd8806 is one:

https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/led_strip/lpd880x.c#L127

Cheap, low pin count devices sometimes don't have a CS pin.

**Comment by mbolivar** [comment]
> I tested it both on 96b_carbon and disco_l475_iot1.

Cool, thanks for confirming.

Just as an FYI, I'm actually moving away from using Carbon, and I wasn't be putting in as much stm32 work in the near future as I thought there were issues preventing IPSP from working properly on disco_l475_iot1. But from your statements, this may be working? Or was it BT testing without IPSP? (I'm out of town and don't have either of those boards with me)

**Comment by erwango** [comment]
> there were issues preventing IPSP from working properly on disco_l475_iot1. 

@mbolivar : We had a doubt on requested BT supported version. But it should actually work properly. I'll spend some time next week on it and I should be able to get some help from dedicated ST team. Hopefully this will help to unlock this scenario.

**Comment by ydamigos** [comment]
> Just as an FYI, I'm actually moving away from using Carbon, and I wasn't be putting in as much stm32 work in the near future as I thought there were issues preventing IPSP from working properly on disco_l475_iot1. But from your statements, this may be working? Or was it BT testing without IPSP? (I'm out of town and don't have either of those boards with me)

I tested both boards using `samples/bluetooth/beacon`.

**Comment by mbolivar** [comment]
Thanks @erwango!

@ydamigos, I think basic beacon support should be enough to verify the SPI changes in this PR are OK. Thanks again for clarifying.

File: boards/arm/96b_carbon/96b_carbon.dts
Comment (galak) [inline boards/arm/96b_carbon/96b_carbon.dts]: The compatible here isn't correct.  Saying "nordic,nrf51822-qfac" isn't specific enough.  We need something that conveys the programming interface/driver one would select for the "device" connected.  In practice, this seems like it would instead be something like "zephyr,bt-spi" (but that's just a guess).

**Review by galak — CHANGES_REQUESTED** [review]
Not sure if BT will be the easiest SPI device to start with.  I'm happy to hash through what we need to get this correct, but need to understand a bit more about what the "actual" devices are in the disco and carbon cases.

In disco, I'm guessing the "bt" device is running some firmware from ST.  I assume in the carbon case its zephyr firmware (or something like that)?

**Comment by mbolivar** [comment]
> I assume in the carbon case its zephyr firmware (or something like that)?

Correct: http://docs.zephyrproject.org/boards/arm/96b_carbon_nrf51/doc/96b_carbon_nrf51.html

File: boards/arm/96b_carbon/96b_carbon.dts
Comment (mbolivar) [reply boards/arm/96b_carbon/96b_carbon.dts]: In case it's relevant, this is the sample implementing the SPI protocol (as used on carbon): http://docs.zephyrproject.org/samples/bluetooth/hci_spi/README.html

**Comment by nzmichaelh** [comment]
(FYI, I'll switch the Arduino Zero SPI driver to this once this PR lands)

**Comment by ydamigos** [comment]
> Not sure if BT will be the easiest SPI device to start with. I'm happy to hash through what we need to get this correct, but need to understand a bit more about what the "actual" devices are in the disco and carbon cases.

I agree that BT isn't the most "straightforward" SPI devices  to add to DT but we will have to add it at some point, so it is reasonable to spend some time to get everything correct before merging.

> In disco, I'm guessing the "bt" device is running some firmware from ST. I assume in the carbon case its zephyr firmware (or something like that)?

Yes, disco_l475_iot1 has ST's SPBTLE-RF module with vendor's firmware, 96b_carbon runs Zephyr's bluetooth-hci-spi driver.

File: boards/arm/96b_carbon/96b_carbon.dts
Comment (ydamigos) [reply boards/arm/96b_carbon/96b_carbon.dts]: I was very skeptical when I added nordic,nrf51822-qfac.yaml file because it is not only a bluetooth module. I agree we need to change it to zephyr,bt-hci-spi.yaml, it would make clear which driver is used when someone reads boards dts file. I will do the changes.

**Comment by ydamigos** [comment]
I added the DT bluetooth binding for "zephyr,bt-hci-spi".

**Comment by ydamigos** [comment]
> I added the DT bluetooth binding for "zephyr,bt-hci-spi".

I added back the DT bluetooth binding for "nordic,nrf51822-qfac" and used compatible to select "zephyr,bt-hci-spi" driver.

File: boards/arm/96b_carbon/96b_carbon.dts
Comment (ydamigos) [reply boards/arm/96b_carbon/96b_carbon.dts]: > I was very skeptical when I added nordic,nrf51822-qfac.yaml file because it is not only a bluetooth module. I agree we need to change it to zephyr,bt-hci-spi.yaml, it would make clear which driver is used when someone reads boards dts file. I will do the changes.

I was not satisfied by the above approach because nrf51822-qfac wasn't mentioned anywhere in the device tree file of 96b_carbon board. I moved back to nordic,nrf51822-qfac.yaml and used compatible property to specify the driver used by the module.

**Review by galak — CHANGES_REQUESTED** [review]
Need to cleanup yaml

File: boards/arm/96b_carbon/96b_carbon.dts
Comment (ydamigos) [reply boards/arm/96b_carbon/96b_carbon.dts]: Ack

File: dts/bindings/bluetooth/nordic,nrf51822-qfac.yaml
Comment (galak) [reply dts/bindings/bluetooth/nordic,nrf51822-qfac.yaml]: Correct, from discussion with the BT zephyr guys they felt that "zephyr,bt-hci-spi" is sufficient.

File: boards/arm/96b_carbon/96b_carbon.dts
Comment (ydamigos) [reply boards/arm/96b_carbon/96b_carbon.dts]: Ok

File: dts/bindings/bluetooth/nordic,nrf51822-qfac.yaml
Comment (ydamigos) [reply dts/bindings/bluetooth/nordic,nrf51822-qfac.yaml]: I will rename the yaml to zephyr,bt-hci-spi.

**Comment by ydamigos** [comment]
@galak I made the changes we discussed.

**Review by galak — APPROVED** [review]

**Pull request merged**
